File: kotlin/풀었던 문제/1780_종이의 개수.kt
var graph : Array<IntArray> = arrayOf()
var minusCount = 0
var zeroCount = 0
var plusCount = 0

fun check(row: Int, col: Int, n: Int):Boolean {
    var start = graph[row][col]

    for(i in row until row + n){
        for(j in col until col + n){
            if(start != graph[i][j]) return false
        }
    }
    return true
}

fun divide(row: Int, col: Int, n: Int){

    if (check(row, col, n)) {
        var start = graph[row][col]

        when(start){
            -1 -> minusCount++
            0 -> zeroCount++
            1 -> plusCount++
        }
    } else {
        var size = n / 3

        for (i in 0 until 3){
            for(j in 0 until 3){
                divide(row + size * i, col + size * j, size)
            }
        }
    }

}

fun main(){
    val br = System.`in`.bufferedReader()
    val bw = System.`out`.bufferedWriter()

    val N = br.readLine().toInt()

    graph = Array(N){ IntArray(N) }

    for(i in 0..graph.lastIndex){
        graph[i] = br.readLine().split(" ").map{ it.toInt() }.toIntArray()
    }

    divide(0, 0, N)

    bw.append("$minusCount\n")
    bw.append("$zeroCount\n")
    bw.append("$plusCount")
    bw.flush()
    bw.close()
}

File: Algorithm For Final Swift/Algorithm For Final Swift/BarkingDog/lecture 11 재귀/1780_종이의 개수.swift
//
//  1780_종이의 개수.swift
//  Algorithm For Final Swift
//
//  Created by Kim Yewon on 2023/06/14.
//

import Foundation

func check(_ n: Int, _ row: Int, _ col: Int) -> Bool {
    let standard = board[row][col]
    
    for i in row ..< row + n {
        for j in col ..< col + n {
            if board[i][j] != standard { return false }
        }
    }
    
    return true
}

func divide(n: Int, row: Int, col: Int) {
    if check(n, row, col) {
        switch board[row][col] {
        case -1: minusCount += 1
        case 0: zeroCount += 1
        case 1: plusCount += 1
        default: break
        }
    } else {
        let size = n / 3
        
        for i in 0 ..< 3 {
            for j in 0 ..< 3 {
                divide(n: size, row: row + (size * i), col: col + (size * j))
            }
        }
    }
}

let n = Int(readLine()!)!

var minusCount = 0
var zeroCount = 0
var plusCount = 0

var board = [[Int]]()
for _ in 0 ..< n {
    board.append(readLine()!.split { $0 == " " }.map { Int($0)! })
}

divide(n: n, row: 0, col: 0)
print("\(minusCount)\n\(zeroCount)\n\(plusCount)")
